### ISSUE TITLE
Gemspec files ignore if installing from `path` or `git`

### ISSUE THREAD

**Issue by glebm** [bundler/bundler#5163]
When a `gem` is specified via `path` or `git` in the Gemfile, bundler will ignore the `files` field of the specification.

This results in certain files being loadable when they shouldn't be.

**Comment by segiddins** [comment]
Please share the information requested in https://github.com/bundler/bundler/blob/master/ISSUES.md#reporting-unresolved-problems so we can investigate. Thanks!

**Comment by glebm** [comment]
I'm using the latest bundler release (1.13.6), RVM and ruby 2.3.1.

I hesitate to provide a script because there is a non-trivial chance of this being closed as Works-as-intended. Perhaps if you confirm that this is a bug?

Here is an explanation with examples:

If you have a Gemfile that depends on another gem by path, e.g.:

```ruby
# Gemfile
gem 'x', path: '/tmp/x'
```

And the `x.gemspec` specifies some files:

```ruby
# x.gemspec
Gem::Specification.new do |s|
  s.gemspec = ['y.rb']
end
```

And `/tmp/x` contains a file not specified by `x.gemspec`, e.g. `x.rb`, that `x.rb` will be available to the script using Gemfile.

This means depending on the gem via `path` or `git` is not the same as depending on the gem *built* from that path or git repo, which is IMO unexpected.

**Comment by segiddins** [comment]
Thinking it over, there's no way around this -- bundler just sets the load path, for it to truly follow the `files` attribute, it would have to delete all other files in your gems directory

**Comment by glebm** [comment]
@segiddins Perhaps it can build the gem and use that, instead of using the files directly? That's kind of what I was expecting.

**Comment by segiddins** [comment]
An RFC for that feature would be required, but I don't think this will be high-priority for anyone on the core team, sorry.

**Comment by agibralter** [comment]
Are there any plans to have bundler "package" and install git/path gems differently? I feel like the expected behavior is for a git/path gem to be installed into the GEM_HOME just as a normal gem. And if you're bundle-packaging, a `my-gem-SHA.gem` to be created in (e.g.) `vendor/cache`

**Comment by indirect** [comment]
@agibralter no, there are no plans to do that. gems in git repos are frequently not able to successfully build into .gem files, and so packaging requires checking out and packaging the entire state of the repo at the sha that the bundle is locked to.
